Generalize `TrieOrChunk` into `ValueOrChunk`

goral09 opened the pull request:
In this PR we're generalizing `TrieOrChunk` which was a "wrapper" for `Bytes` into `ValueOrChunk` which allows for reusing the chunking mechanism for other types.

After that, I've remove the `get_trie` method from `Storage` component - it did the chunking which isn't the "reponsibility" of storage/execution_engine. Now we do the chunking only in `contract_runtime` which seems more appropriate. This opens up a possibility to cache the trie (other types) rather than throw everything away after it's chunked.

Next, we extract `ValueOrChunk` and `TrieOrChunk` out of `execution_engine` and move it to `casper_node`. 

TODO:
- [x] Finish writing new tests for `ValueOrChunk` that are going to replace the previous tests from `execution_engine` (which are now commented out).

Commit:
Rename TrieOrChunk enum to ValueOrChunk.

Commit:
Update .gitignore file

Commit:
Introduce TrieRaw newtype and return it from get_trie_full

Commit:
Remove Storage::get_trie and do the chunking in ContractRuntime

Commit:
Move ValueOrChunk from execution_engine out to casper_node::types

Commit:
Remove unused error variant from the EE and put in ContractRuntime.

Commit:
Add a chunking constructor and rewrite tests.

Fraser999 commented on node/src/components/contract_runtime.rs:
```diff
@@ -858,12 +862,25 @@ impl ContractRuntime {
-        let result = engine_state.get_trie(correlation_id, trie_or_chunk_id);
-        metrics.get_trie.observe(start.elapsed().as_secs_f64());
-        result
+        let TrieOrChunkId(trie_index, trie_key) = trie_or_chunk_id;
```
```suggestion
        let TrieOrChunkId(chunk_index, trie_key) = trie_or_chunk_id;
```

goral09 replied:
Done in [`3da4c1b` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/3da4c1be1ed4e3584136bf18848f53685a90e1df)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+    where
+        V: ToBytes,
+    {
+        let bytes = ToBytes::into_bytes(data).map_err(ChunkingError::SerializationError)?;
```
The `AsRef<u8>` approach didn't pan out?

goral09 replied:
Yes, for a couple of reasons: 
1. That would require having implemented `AsRef<[u8]>` for all types with which me might want to call this. We will be, most likely, calling `ValueOrChunk::new` with types that we already transfer over the wire - block effects, deploy (approvals), tries, etc. 
2. The choice of serialization (or turning `T` into bytes) has to be aligned with how it's deserialized in the fetcher since its hash might be used for id calculation/proof verification.

It seemed to me that using `ToBytes` will be the safest and most reasonable approach.

goral09 replied:
Actually, now that I've learned that we use `bincode` to serialize data before sending, we should also use that same serialization mechanism for turning `T` into `Vec<u8>` - after all, we want to learn how big is the _wire representation_ of `T`.

goral09 replied:
Done in [`4554498` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/455449814e93066bccd477bde49499152157ac94)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// Represents an enum that can contain either a whole trie or a chunk of it.
+pub type TrieOrChunk = ValueOrChunk<Bytes>;
```
This will become `ValueOrChunk<TrieRaw>`, yes?

goral09 replied:
Done [`c550302` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/c550302f242e9ca23d65404e38bf1aab0b09bff7)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+    }
+}
+
+impl<V: ToBytes> ToBytes for ValueOrChunk<V> {
```
We need bytesrepr roundtrip tests for this type.

goral09 replied:
I've completely removed `ToBytes` and `FromBytes` impls for `ValueOrChunk` since it was unused - [`512e06c` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/512e06c27f01cc8cbfeb2bb986ee33922400581a).

Requests/Responses are serialized using `bincode`:

1. Construct a get response - [code](https://github.com/casper-network/casper-node/blob/dev/node/src/components/contract_runtime.rs#L260)
2. Calls to `Message::new_get_response` call `Item::to_serialized` - [code](https://github.com/casper-network/casper-node/blob/dev/node/src/protocol.rs#L170)
3. `to_serialized` calls `bincode::serialize` - [code](https://github.com/casper-network/casper-node/blob/dev/node/src/components/fetcher.rs#L94)`

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+/// The index is the index of a chunk if the node's size is too large and requires chunking. For
+/// small nodes, it's always 0.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkId(pub u64, pub Digest);
```
We generally use named fields for structs with more than one field - I'd certainly prefer that here.

goral09 replied:
This is just moved over from the `execution_engine/trie.rs`.

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+    }
+
+    /// Given a serialized ID, deserializes it for display purposes.
+    pub fn fmt_serialized(f: &mut Formatter, serialized_id: &[u8]) -> fmt::Result {
```
```suggestion
    fn fmt_serialized(f: &mut Formatter, serialized_id: &[u8]) -> fmt::Result {
```

goral09 replied:
Didn't change this code either. Just copy-paste.

goral09 replied:
Done in [`3da4c1b` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/3da4c1be1ed4e3584136bf18848f53685a90e1df)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+}
+
+/// Helper struct to on-demand deserialize a trie or chunk ID for display purposes.
+pub struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
```
```suggestion
pub(crate) struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
```

goral09 replied:
Done in [`3da4c1b` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/3da4c1be1ed4e3584136bf18848f53685a90e1df)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,207 @@
+            Ok(ValueOrChunk::Value {
+                data: bytes.into(),
+                _phantom: PhantomData,
+            })
```
Should we print a warning and return an error here if `chunk_index != 0`?  Seems like that would be the correct thing to do in case we were handling a malformed/malicious request maybe?

goral09 replied:
We will return it inside the [`ChinkWithProof::new`](https://github.com/casper-network/casper-node/blob/dev/hashing/src/chunk_with_proof.rs#L79)

Commit:
Use TrieRaw type for TrieOrChunk (instead of Bytes).

Commit:
Change visiblity of struct and method.

Commit:
Use hardcoded CHUNK_SIZE from ChunkWithProof.

Commit:
Extend ValueOrChunk tests.

Commit:
Remove unused ToBytes/FromBytes for ValueOrChunk.

Commit:
Remove commented out tests for `get_trie`.

The original test was written against the API of the storage and
recently moved to contract_runtime.rs during extraction of TrieOrChunk.
That test queried trie storage to test that when data gets chunked it
can be restored to its original form later. It set up the trie in such a
way the structure of the trie led to a leaf exceeded the chunk size limit.
After generalizing (and extracting) it to ValueOrChunk and moving
chunking to the constructor, we can test it in isolation without the
special setup of the trie. That test exists in value_or_chunk.rs file.

Commit:
Add newline at the end of .gitignore file.

Commit:
Replace type bound on V in ValueOrChunk constructor.

We want to chunk when the serialized representation of a type exceeds
our limit. When determining byte representation we should use the same
method that we will use when we actually serialize the data for sending.
Currently, we use bincode::serialize (see Message::new_get_response) so
we should also use it when we chunk.

rafal-ch commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of `ValueOrChunk::Value` if data is less than `max_chunk_size`
```
```suggestion
    /// Creates an instance of `ValueOrChunk::Value` if data size is less than or equal to [ChunkWithProof::CHUNK_SIZE_BYTES]
```

Fraser999 replied:
Agree with the suggestion but would like the `[ChunkWithProof::CHUNK_SIZE_BYTES]` to include backticks.

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

rafal-ch commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+#[cfg(test)]
+mod value_chunking {
```
I believe this should say `mod tests`.
We could consider introducing a submodule called `value_chunking` to group the tests, but it's something we avoided so far in the `casper-node`. Also, this is just a single test so there's no need for sophisticated grouping :)

goral09 replied:
Why should it say `mod tests`? Is it a convention?

rafal-ch replied:
On the contrary to the requirement for the integration tests folder to be named `tests`, this one is not a strict requirement. It's more like de-facto standard.

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

rafal-ch commented on node/src/components/contract_runtime.rs:
```diff
@@ -858,12 +862,21 @@ impl ContractRuntime {
-        metrics.get_trie.observe(start.elapsed().as_secs_f64());
-        result
+        let TrieOrChunkId(chunk_index, trie_key) = trie_or_chunk_id;
+        match engine_state.get_trie_full(correlation_id, trie_key)? {
+            None => {
+                metrics.get_trie.observe(start.elapsed().as_secs_f64());
+                Ok(None)
+            }
+            Some(trie_raw) => {
+                let trie_or_chunk = TrieOrChunk::new(trie_raw, chunk_index)?;
+                metrics.get_trie.observe(start.elapsed().as_secs_f64());
+                Ok(Some(trie_or_chunk))
+            }
+        }
```
Could this work?
Rationale: try to update the metric in a single place.

``` suggestion
        let ret = match engine_state.get_trie_full(correlation_id, trie_key)? {
            None => Ok(None),
            Some(trie_raw) => Ok(Some(TrieOrChunk::new(trie_raw, chunk_index)?)),
        };
        metrics.get_trie.observe(start.elapsed().as_secs_f64());
        ret
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

rafal-ch reviewed (commented):
LGTM :+1: 

I feel inclined to approve right away, but first I'd like to run the `join_with_big_trie` test from `casper-test` against this branch. I'll check if I can fix the test quickly, because currently it fails due to other changes in the node.

Fraser999 commented on execution_engine/src/storage/trie/mod.rs:
```diff
@@ -349,124 +348,36 @@ impl ::std::fmt::Debug for PointerBlock {
-                .field("hash", &chunk.proof().root_hash())
-                .finish(),
-        }
+impl TrieRaw {
```
I think we only need this `impl` if the inner bytes aren't `pub`.  We generally don't make wrapped types `pub` so I guess I'd rather we took that approach here too.  But if you strongly prefer to keep them `pub` then let's remove this `impl`.

goral09 replied:
No, I think you're right. It might be that I've iteratively got from inner `pub` to an `impl` and forgot to remove the former.

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types.rs:
```diff
@@ -46,6 +47,7 @@ pub use node_config::NodeConfig;
 pub(crate) use node_id::NodeId;
 pub use peers_map::PeersMap;
 pub use status_feed::{ChainspecInfo, GetStatusResult, NodeState, StatusFeed};
+pub use value_or_chunk::*;
```
It's a pain I agree, but I'd rather be explicit about which types and functions are being exported rather than using `*`.

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+use datasize::DataSize;
+
+/// Represents a value or a chunk of data with attached proof.
+/// Chunk with attached proof is used when the requested
+/// value is larger than [ChunkWithProof::CHUNK_SIZE_BYTES].
```
```suggestion
///
/// Chunk with attached proof is used when the requested
/// value is larger than [`ChunkWithProof::CHUNK_SIZE_BYTES`].
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+}
+
+#[derive(Debug, Error)]
+/// Error returned when constructing an instance of `ValueOrChunk`.
```
```suggestion
/// Error returned when constructing an instance of [`ValueOrChunk`].
#[derive(Debug, Error)]
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+#[derive(Debug, Error)]
+/// Error returned when constructing an instance of `ValueOrChunk`.
+pub enum ChunkingError {
+    /// Merkle proof construction error
```
```suggestion
    /// Merkle proof construction error.
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+    /// Merkle proof construction error
+    #[error("error constructing merkle proof for chunk")]
+    MerkleConstruction(#[from] MerkleConstructionError),
+    /// Serialization error
```
```suggestion
    /// Serialization error.
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+
+impl<V> ValueOrChunk<V> {
+    /// Creates an instance of `ValueOrChunk::Value` if data is less than `max_chunk_size`
+    /// or a `ValueOrChunk::ChunkWithProof` if it is. In the latter case it will return only the
```
```suggestion
    /// or a `ValueOrChunk::ChunkWithProof` if it is greater. In the latter case it will return only the
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

Fraser999 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,164 @@
+    use super::ValueOrChunk;
+
+    fn create_test_bytes(length: usize) -> Bytes {
+        std::iter::repeat(1u8)
+            .take(length)
+            .collect::<Vec<u8>>()
+            .into()
```
```suggestion
        vec![1_u8; length].into()
```

goral09 replied:
Done in [`e8243e8` (#3255)](https://github.com/casper-network/casper-node/pull/3255/commits/e8243e8b08580a6e72cebdfed0b1b1957f46652a)

rafal-ch reviewed (changes requested):
> but first I'd like to run the join_with_big_trie test from casper-test against this branch

Blocking temporarily, as I can't make the test pass on this branch. I'll follow up with the details.

Commit:
Address review comments.

Commit:
Use ToBytes impl, instead of bincode, when chunking.

When we write `Trie` to trie store, or when we calculate hash of the
trie, we use `ToBytes` impls to go from the trie to bytes, before
calling [`Digest::hash_bytes_into_chunks_if_necessary`] on the bytes. If
we used different serialization (type to bytes conversion) when
chunking, the proofs in the chunks would not hash to the merkle tree of
the serialized trie from the trie store == different bytes will lead to
different merkle proofs.

This alignment needs to be maintained for all types that want to be
chunked and verified against merkle proofs - the same serialization
mapping needs to be used when calculating merkle root hash and when
chunking.

It seems the easiest to use `bytesrepr::ToBytes` implementations
everywhere, rather than making this generic and endforcing with
scattered, e2e tests.

fizyk20 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,172 @@
+        // NOTE: Cannot accept the chunk size bytes as an argument without changing the
+        // IndexedMerkleProof. The chunk size there is hardcoded and will be used when
+        // determining the chunk.
+        if bytes.len() <= ChunkWithProof::CHUNK_SIZE_BYTES {
```
Should we require that `chunk_index` is 0 if that's the case?

goral09 replied:
And fail if it's different? Current approach seems to be to be the most idiot-proof: even if someone passes in an invalid `chunk_index` they will get a correct results. Especially that these chunk sizes are hardcoded so expecting a `chunk_index != 0` where the size of the `bytes` is less than the limit is wrong.

fizyk20 replied:
> Especially that these chunk sizes are hardcoded so expecting a chunk_index != 0 where the size of the bytes is less than the limit is wrong.

Exactly, which is why returning `Ok` when the chunk index doesn't make sense might hide bugs in other pieces of code. Like if someone writes a loop that goes over multiple indices without checking what the actual count is - they will be under the impression that they got multiple chunks even though they only got one, but repeated multiple times.
I'm tempted to say that this might also potentially be a security issue, but I can't come up with any examples of how it would be, so maybe it isn't. Just feels a bit risky somehow.

fizyk20 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,172 @@
+/// The index is the index of a chunk if the node's size is too large and requires chunking. For
+/// small nodes, it's always 0.
+#[derive(DataSize, Debug, Clone, Copy, Serialize, Deserialize, PartialEq, Eq, Hash)]
+pub struct TrieOrChunkId(pub u64, pub Digest);
```
Wouldn't such ID be exactly the same for `ValueOrChunk<V>` in general? If so, maybe it would make sense to rename it to `ValueOrChunkId`?

goral09 replied:
I didn't want to optimize prematurely and it turned out to be a good decision - I've already done some work on chunking of `Vec<ExecutionResult>` and there we will have different IDs for pre-1.5 and post-1.5 b/c we can't identify the values just based on the data.

fizyk20 commented on node/src/types/value_or_chunk.rs:
```diff
@@ -0,0 +1,172 @@
+}
+
+/// Helper struct to on-demand deserialize a trie or chunk ID for display purposes.
+pub struct TrieOrChunkIdDisplay<'a>(pub &'a [u8]);
```
Same as for `TrieOrChunkId`.

fizyk20 replied:
Resolving because there apparently is a good reason for not doing that (https://github.com/casper-network/casper-node/pull/3255#discussion_r953816573)

fizyk20 reviewed (approved):
Nice! I only left a few non-blocking comments.

Commit:
Add a test for fetchin big trie from the ContractRuntime.

Commit:
Add  trait.

Commit:
Rename hash_bytes_into_chunks_if_necessary

rafal-ch commented on hashing/src/chunk_with_proof.rs:
```diff
@@ -122,6 +124,48 @@ impl ChunkWithProof {
     }
 }
 
+/// Implemented for types that are chunked when sending over the wire and/or before storing the the
```
```suggestion
/// Implemented for types that are chunked when sending over the wire and/or before storing the
```

rafal-ch commented:
> Blocking temporarily, as I can't make the test pass on this branch.

The `join_with_big_trie` test passes when using commit 03192a305b86890322b9a0a8c54bdb09cbfee43f :+1:

rafal-ch reviewed (approved):
LGTM :+1: 

Approving with one minor nit (typo).

Commit:
Fix docs.

Commit:
Fix unit test.

Commit:
Move Chunkable from hashing to casper-node.

Commit:
Change return type of Chunkable::as_bytes

Commit:
Merge remote-tracking branch 'upstream/dev' into generalize_trie_or_chunk

Commit:
Fix docs.

Commit:
Fix formatting.
